### PR TITLE
Fetch network info and parse it

### PR DESCRIPTION
**Pull request by biafra23**
Handles #1530 

Changes proposed in this pull request:
- Introduce `ChainInfoRepository`
- Fetch from new endpoint (`v1/chains`) and parse JSON
- Modify all other calls to previous endpoints to have a default chaind of 4 (Rinkeby)

@gnosis/mobile-devs

### CODE REVIEW

File: data/src/main/java/io/gnosis/data/backend/GatewayApi.kt [hunk 1]
```diff
@@ -14,43 +15,71 @@ import java.util.*
 
 interface GatewayApi {
 
-    @GET("/v1/balances/supported-fiat-codes")
-    suspend fun loadSupportedCurrencies(): List<String>
+    @GET("/{chainId}/v1/balances/supported-fiat-codes")
+    suspend fun loadSupportedCurrencies(
+        @Path("chainId") chainId: Int = 4
```
Comment (elgatovital) [inline data/src/main/java/io/gnosis/data/backend/GatewayApi.kt:20]: default value could come from build constants (BuildConfig.CHAIN_ID). I think we even have it already

File: data/src/main/java/io/gnosis/data/db/daos/ChainDao.kt [hunk 2]
```diff
@@ -0,0 +1,27 @@
+package io.gnosis.data.db.daos
+
+import androidx.room.*
+import io.gnosis.data.models.Chain
+
+@Dao
+interface ChainDao {
+
+    @Insert(onConflict = OnConflictStrategy.REPLACE)
+    suspend fun insert(chain: Chain)
```
Comment (elgatovital) [inline data/src/main/java/io/gnosis/data/db/daos/ChainDao.kt:10]: insert -> save as we will probably use this to also update the chain info

File: data/src/main/java/io/gnosis/data/models/Chain.kt [hunk 3]
```diff
@@ -0,0 +1,24 @@
+package io.gnosis.data.models
+
+import androidx.room.ColumnInfo
+import androidx.room.Entity
+import androidx.room.PrimaryKey
+import io.gnosis.data.models.Chain.Companion.TABLE_NAME
+
+@Entity(tableName = TABLE_NAME)
+data class Chain(
```
Comment (elgatovital) [inline data/src/main/java/io/gnosis/data/models/Chain.kt:9]: we could combine Chain & ChainInfo into one class as we currently do not have separate dto's and use models from cgw directly

File: data/src/main/java/io/gnosis/data/models/ChainInfo.kt [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+
+@JsonClass(generateAdapter = true)
+data class ChainInfo(
+    @Json(name = "chainId") val chainId: Int,
+    @Json(name = "chainName") val chainName: String,
+    @Json(name = "rpcUrl") val rpcUrl: String,
+    @Json(name = "blockExplorerUrl") val blockExplorerUrl: String,
+    @Json(name = "nativeCurrency") val nativeCurrency: NativeCurrency,
+    @Json(name = "transactionService") val transactionService: String,
```
Comment (elgatovital) [inline data/src/main/java/io/gnosis/data/models/ChainInfo.kt:14]: we need transactionService url only to possibly show it on network details page?

File: data/src/main/java/io/gnosis/data/repositories/ChainInfoRepository.kt [hunk 5]
```diff
@@ -0,0 +1,28 @@
+package io.gnosis.data.repositories
+
+import io.gnosis.data.backend.GatewayApi
+import io.gnosis.data.db.daos.ChainDao
+import io.gnosis.data.models.Chain
+import io.gnosis.data.models.ChainInfo
+
+class ChainInfoRepository(
```
Comment (elgatovital) [inline data/src/main/java/io/gnosis/data/repositories/ChainInfoRepository.kt:8]: how about putting chain related stuff into SafeRepository?

File: data/src/main/java/io/gnosis/data/models/ChainInfo.kt (hunk 4 above)
Comment (biafra23) [reply data/src/main/java/io/gnosis/data/models/ChainInfo.kt:14]: For the DTO model I wanted to include everything we get. No matter if we need it. I would reduce the data we store to what we need now or until the feature is delivered.

File: data/src/main/java/io/gnosis/data/repositories/ChainInfoRepository.kt (hunk 5 above)
Comment (biafra23) [reply data/src/main/java/io/gnosis/data/repositories/ChainInfoRepository.kt:8]: I put it there at first. But there might be parts of the app that want to get network info without using a safe.

File: data/src/main/java/io/gnosis/data/models/Chain.kt (hunk 3 above)
Comment (biafra23) [reply data/src/main/java/io/gnosis/data/models/Chain.kt:9]: How would we handle embedded types like NativeCurrency and ChainTheme? Flatten them into the chain table or have relations?

Comment (elgatovital) [reply data/src/main/java/io/gnosis/data/models/Chain.kt:9]: flatten them

**Review by elgatovital — APPROVED** [review]

**Pull request merged**